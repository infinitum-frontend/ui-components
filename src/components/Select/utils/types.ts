import { InputProps } from 'Components/Input'
import { PopoverContentProps, PopoverProps } from 'Components/Popover'
import { ComponentPropsWithoutRef, ReactNode } from 'react'

export interface SelectProps<Multiple extends boolean = false>
  extends Omit<
    ComponentPropsWithoutRef<'button'>,
    'value' | 'prefix' | 'onChange'
  > {
  /** Активация множественного выбора */
  multiple?: Multiple
  /** Список опций. Может быть сгруппирован */
  options?: SelectOptions
  /** Controlled value */
  value?: Multiple extends true ? SelectValue[] : SelectValue | undefined
  /** Controlled onChange */
  onChange?: (
    value: Multiple extends true ? SelectOption[] : SelectOption
  ) => void
  /** Обработчик нажатия на кнопку очистки значения, которая отображается при clearable. Можно определить в нём произвольную логику. Если его не передать, то будет вызван onChange */
  onClear?: () => void
  /** Callback на изменение значения поля фильтрации. Если true, то дефолтная фильтрация не используется, ожидается, что снаружи будет передан отфильтрованный options */
  onFilterChange?: (filterValue: string) => void
  /** Отображение поля фильтрации опций */
  filterable?: boolean
  /** Отображение кнопки очистки выбранного значения. При нажатии на кнопку вызывается обработчик onClear, а если он не был передан, то onChange. */
  clearable?: boolean
  /** Состояние недоступности */
  disabled?: boolean
  /** Обязательно к заполнению */
  required?: boolean
  /** Состояние загрузки */
  loading?: boolean
  /** Loader может быть расположен в SelectButton или во всплывающем окне */
  loaderPlacement?: 'inline' | 'dropdown'
  /** Placeholder, если нет выбранного значения */
  placeholder?: string
  /** Размер SelectButton */
  size?: InputProps['size']
  /** Слот для размещения элемента (иконки) слева */
  prefix?: ReactNode
  /** Ширина всплывающего окна */
  popoverWidth?: PopoverContentProps['width']
  /** Расположение всплывающего окна относительно SelectButton */
  popoverPlacement?: PopoverProps['placement']
  /** Использование кастомного шаблона SelectButton */
  renderControl?: (props: {
    displayValue: string
    isOpen: boolean
    disabled: boolean
    required: boolean
    ref: React.ForwardedRef<HTMLButtonElement>
    onClick: React.MouseEventHandler<HTMLButtonElement> | undefined
  }) => React.ReactElement
  /** Поле фильтрации может быть расположено в SelectButton или во всплывающем окне */
  filterPlacement?: 'inline' | 'dropdown'
  /** Дополнительный className */
  className?: string
  /** Максимальное количество отображаемых элементов, после которого начинается скролл
   * @deprecated use maxHeight
   */
  maxItemsCount?: number
  /**
   * Максимальная высота контента, после которой начинается скролл
   */
  maxHeight?: number
  /** Виртуализация списка опций */
  virtualized?: boolean
  /** Текст сообщения, если нет опций */
  emptyMessage?: string
  /** Текст подсказки во всплывающем окне */
  dropdownHint?: string
  // ==================================== Нужны ли пропы ниже непонятно
  /** Состояние валидности */
  status?: 'error'
  // autoFocus?: boolean  /** Кеширование асинхронно загруженных опций */
  // cacheOptions?: boolean
  /** Показывать выбранные значения в начале списка */
  // selectedFirst?: boolean
  // loadOptions?: (filterValue?: string) => Promise<SelectOptions>
  // cacheOptions?: boolean
  // filterOptions?: (options: SelectOptions) => SelectOptions
}

export type SelectValue = string | number

export interface DefaultSelectOption {
  value: SelectValue
  label: string
}

type MappedObject<T> = {
  [Property in keyof T]: T[Property]
}

export type SelectOption<T = Record<string, any>> = DefaultSelectOption &
  MappedObject<T>

// export interface SelectOption {
//   value: SelectValue
//   label: string
// }

export interface SelectOptionGroup {
  label: string
  options: SelectOption[]
}

export type SelectOptions = Array<SelectOption | SelectOptionGroup>

export interface GroupLabelItem {
  groupLabel: string
}
export type FlattenOption = SelectOption | GroupLabelItem

export interface UseSelectProps<Multiple extends boolean = false> {
  multiple?: Multiple
  value?: Multiple extends true ? SelectValue[] : SelectValue | undefined
  onChange?: (
    value: Multiple extends true ? SelectOption[] : SelectOption
  ) => void
  options: FlattenOption[]
  onClear?: SelectProps['onClear']
}

export interface UseSelectResult {
  selectedOptions: SelectOption[]
  handleSelect: (option: SelectOption) => void
  handleClear: () => void
  checkOptionSelection: (option: SelectOption) => boolean
  hasSelectedValue: boolean
  isOpen: boolean
  setOpen: (value: boolean) => void
  displayValue: string
}

export interface UseSelectOptionsProps {
  options: SelectOptions
  filterable: SelectProps['filterable']
  filterValue: string
  customFiltering: boolean
}

export interface UseSelectOptionsResult {
  filteredOptions: SelectOptions
  flattenOptions: FlattenOption[]
  filteredFlattenOptions: FlattenOption[]
}

export interface SelectDataFormatterOptions<T extends Record<string, any>> {
  /**
   * Исходный массив, подлежащий форматированию
   */
  array: T[]
  /**
   * Строковое название ключа элемента исходного массива, на основании которого будет браться ключ для опции
   * Ключ должен являться уникальным значением
   */
  value: keyof T
  /**
   * Строковое название ключа элемента исходного массива, на основании которого будет браться текст для отображения опции
   */
  label: keyof T
}
