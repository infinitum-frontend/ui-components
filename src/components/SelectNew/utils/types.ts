import { InputProps } from 'Components/Input'

export interface SelectProps<Multiple extends boolean = false> {
  multiple?: Multiple
  options: SelectOptions
  value: Multiple extends true ? SelectValue[] : SelectValue | undefined
  onChange: (
    value: Multiple extends true ? SelectOption[] : SelectOption
  ) => void
  onClear?: () => void
  loadOptions?: (filterValue?: string) => Promise<SelectOptions>
  filterable?: boolean
  clearable?: boolean
  disabled?: boolean
  loading?: boolean
  placeholder?: string
  size?: InputProps['size']
  // TBD
  filterPlacement?: 'inline' | 'dropdown'
  cacheOptions?: boolean
  className?: string
}

export type SelectValue = string | number

export interface SelectOption {
  value: SelectValue
  label: string
}

export interface SelectOptionGroup {
  label: string
  options: SelectOption[]
}

export type SelectOptions = Array<SelectOption | SelectOptionGroup>

export interface GroupLabelItem {
  groupLabel: string
}
export type FlattenOption = SelectOption | GroupLabelItem

export interface UseSelectProps<Multiple extends boolean = false> {
  multiple?: Multiple
  value: Multiple extends true ? SelectValue[] : SelectValue | undefined
  onChange: (
    value: Multiple extends true ? SelectOption[] : SelectOption
  ) => void
  options: FlattenOption[]
  onClear?: SelectProps['onClear']
}

export interface UseSelectResult {
  selectedOptions: SelectOption[]
  handleSelect: (option: SelectOption) => void
  handleClear: () => void
  checkOptionSelection: (option: SelectOption) => boolean
  hasSelectedValue: boolean
  isOpen: boolean
  setOpen: (value: boolean) => void
  displayValue: string
}

export interface UseSelectOptionsProps {
  options: SelectOptions
  filterable: SelectProps['filterable']
  filterValue: string
}

export interface UseSelectOptionsResult {
  filteredOptions: SelectOptions
  flattenOptions: FlattenOption[]
  filteredFlattenOptions: FlattenOption[]
}
