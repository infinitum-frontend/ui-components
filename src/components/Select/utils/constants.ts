import { SelectOption } from './types'

export const SELECT_DROPDOWN_SELECTOR = 'inf-select-dropdown'

export const defaultSelectItem: SelectOption = {
  value: -1, // TODO: null?
  label: 'Не указано'
}

export const SELECT_OPTION_HEIGHT = 36
