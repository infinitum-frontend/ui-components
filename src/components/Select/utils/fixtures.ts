// export const SelectBaseOptions = [
//   { label: 'Депозитарные услуги', value: 'one' },
//   { label: 'Очень длинные спецдепозитарные услуги', value: 'two' },
//   { label: 'Консалтинг и аутсорсинг', value: 'three' }
// ]

export const SelectBaseOptions = [
  { label: 'Депозитарные услуги', value: 0 },
  { label: 'Спецдепозитарные услуги', value: 1 },
  { label: 'Консалтинг и аутсорсинг', value: 2 }
]

export const SelectBaseGroupedOptions = [
  { label: 'Депозитарные услуги', value: 'one' },
  {
    label: 'Спецдепозитарные услуги',
    options: [
      { label: 'Спецдепозитарные услуги 1', value: 'two-one' },
      { label: 'Очень длинные спецдепозитарные услуги 2', value: 'two-two' },
      { label: 'Спецдепозитарные услуги 3', value: 'two-three' }
    ]
  },
  {
    label: 'Консалтинг и аутсорсинг',
    options: [{ label: 'Консалтинг и аутсорсинг 1', value: 'three-one' }]
  }
]

export const SelectLongOptions = [
  { label: 'Депозитарные услуги', value: 'one' },
  { label: 'Очень длинные спецдепозитарные услуги', value: 'two' },
  { label: 'Консалтинг и аутсорсинг', value: 'three' },
  { label: 'Консалтинг и аутсорсинг 2', value: 'four' },
  { label: 'Консалтинг и аутсорсинг 3', value: 'five' },
  { label: 'Консалтинг и аутсорсинг 4', value: 'six' },
  { label: 'Консалтинг и аутсорсинг 5', value: 'sevent' },
  { label: 'Консалтинг и аутсорсинг 6', value: 'eight' },
  { label: 'Консалтинг и аутсорсинг 7', value: 'nine' },
  { label: 'Консалтинг и аутсорсинг 8', value: 'ten' }
]

export const SelectOptionsUnformatted = [
  { text: 'Депозитарные услуги', id: 0 },
  { text: 'Спецдепозитарные услуги', id: 1 },
  { text: 'Консалтинг и аутсорсинг', id: 2 }
]

export const SelectOptionsWithExtraKeys = [
  { text: 'Депозитарные услуги', id: 0, subtext: 'Subtext' },
  { text: 'Спецдепозитарные услуги', id: 1, subtext: 'Subtext' },
  { text: 'Консалтинг и аутсорсинг', id: 2, subtext: 'Subtext' }
]

export const SelectOptionsRawLong = [
  {
    id: '1',
    name: 'Ошибка доставки'
  },
  {
    id: '2',
    name: 'На отправке'
  },
  {
    id: '3',
    name: 'Доставлен'
  },
  {
    id: '4',
    name: 'Получен'
  },
  {
    id: '5',
    name: 'Ошибка в обработке'
  },
  {
    id: '6',
    name: 'Отправлен'
  },
  {
    id: '7',
    name: 'Отказано в обработке'
  },
  {
    id: '8',
    name: 'Сохранено'
  }
]
