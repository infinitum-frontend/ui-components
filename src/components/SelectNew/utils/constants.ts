import { SelectOption } from './types'

export const SELECT_DROPDOWN_SELECTOR = 'inf-select-dropdown'

export const defaultSelectItem: SelectOption = {
  value: -1, // TODO: null?
  label: 'Не указано'
}
