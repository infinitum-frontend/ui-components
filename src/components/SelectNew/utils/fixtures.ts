export const SelectBaseOptions = [
  { label: 'Депозитарные услуги', value: 'one' },
  { label: 'Очень длинные спецдепозитарные услуги', value: 'two' },
  { label: 'Консалтинг и аутсорсинг', value: 'three' }
]

export const SelectBaseGroupedOptions = [
  { label: 'Депозитарные услуги', value: 'one' },
  {
    label: 'Спецдепозитарные услуги',
    options: [
      { label: 'Спецдепозитарные услуги 1', value: 'two-one' },
      { label: 'Очень длинные спецдепозитарные услуги 2', value: 'two-two' },
      { label: 'Спецдепозитарные услуги 3', value: 'two-three' }
    ]
  },
  {
    label: 'Консалтинг и аутсорсинг',
    options: [{ label: 'Консалтинг и аутсорсинг 1', value: 'three-one' }]
  },
  { label: 'Другое', value: 'four' }
]

export const SelectLongOptions = [
  { label: 'Депозитарные услуги', value: 'one' },
  { label: 'Очень длинные спецдепозитарные услуги', value: 'two' },
  { label: 'Консалтинг и аутсорсинг', value: 'three' },
  { label: 'Консалтинг и аутсорсинг 2', value: 'four' },
  { label: 'Консалтинг и аутсорсинг 3', value: 'five' },
  { label: 'Консалтинг и аутсорсинг 4', value: 'six' },
  { label: 'Консалтинг и аутсорсинг 5', value: 'sevent' },
  { label: 'Консалтинг и аутсорсинг 6', value: 'eight' },
  { label: 'Консалтинг и аутсорсинг 7', value: 'nine' },
  { label: 'Консалтинг и аутсорсинг 8', value: 'ten' }
]
