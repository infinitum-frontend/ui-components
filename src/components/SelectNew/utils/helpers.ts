import reactNodeToString from '~/src/utils/helpers'
import {
  FlattenOption,
  SelectOption,
  SelectOptionGroup,
  SelectOptions
} from './types'

export const getFilteredFlattenOptions = (
  options: FlattenOption[],
  filterValue: string
): SelectOption[] => {
  console.log('HEAVY: getFilteredFlattenOptions')
  // @ts-expect-error
  return options.filter((option) => {
    if (isGroupLabel(option)) {
      return false
    }
    // @ts-expect-error
    return filterOptionsFn(option, filterValue)
  })
}

/**
 * Фильтрация списка опций по строке поиска
 */
export const getFilteredGroupedOptions = (
  options: SelectOptions,
  filterValue: string
): SelectOptions => {
  console.log('HEAVY: getFilteredOptions')
  const filtered: SelectOptions = []

  options.forEach((item) => {
    const isGroup = isGroupOption(item)

    if (isGroup) {
      // Фильтруем опции внутри группы
      // @ts-expect-error
      const filteredInGroup = item.options.filter((option) =>
        filterOptionsFn(option, filterValue)
      )
      if (filteredInGroup.length) {
        filtered.push({
          label: item.label,
          options: filteredInGroup
        })
      }
    } else {
      // Фильтруем обычную опцию
      // @ts-expect-error
      if (filterOptionsFn(item, filterValue)) {
        filtered.push(item)
      }
    }
  })

  return filtered
}

const filterOptionsFn = (
  option: SelectOption,
  filterValue: string
): boolean => {
  const label =
    typeof option.label === 'string'
      ? option.label
      : reactNodeToString(option.label)
  return label.toLocaleLowerCase().includes(filterValue.toLocaleLowerCase())
}

export const getFlattenOptions = (options: SelectOptions): FlattenOption[] => {
  console.log('HEAVY: getFlattenOptions')
  const flatList: FlattenOption[] = []

  options.forEach((item) => {
    const isGroup = 'options' in item

    if (isGroup) {
      flatList.push(
        ...[
          {
            groupLabel: item.label
          },
          ...item.options
        ]
      )
    } else {
      flatList.push(item)
    }
  })

  return flatList
}

export const isGroupLabel = (option: FlattenOption): boolean => {
  return 'groupLabel' in option
}

export const isGroupOption = (
  option: SelectOption | SelectOptionGroup
): boolean => {
  return 'options' in option
}

// ======= ЧЕРНОВИК =======

// import { produce } from 'immer'
// import reactNodeToString from '~/src/utils/helpers'
// import { SelectOption, SelectOptions, SelectValue } from '../types'

// /**
//  * Фильтрация списка опций по строке поиска
//  */
// export const filterOptions = (
//   options: SelectOptions,
//   filterValue: string
// ): SelectOptions => {
//   const filtered: SelectOptions = []

//   options.forEach((item) => {
//     const isGroup = 'options' in item

//     if (isGroup) {
//       // Фильтруем опции внутри группы
//       const filteredInGroup = item.options.filter((option) =>
//         filterOptionsFn(option, filterValue)
//       )
//       if (filteredInGroup.length) {
//         filtered.push({
//           label: item.label,
//           options: filteredInGroup
//         })
//       }
//     } else {
//       // Фильтруем обычную опцию
//       if (filterOptionsFn(item, filterValue)) {
//         filtered.push(item)
//       }
//     }
//   })

//   return filtered

//   // return produce(options, (draft) => {
//   //   draft = draft.filter((item) => {
//   //     const isGroup = 'options' in item

//   //     if (isGroup) {
//   //       // Фильтруем опции внутри группы
//   //       item.options = item.options.filter((option) =>
//   //         filterOptionsFn(option, filterValue)
//   //       )
//   //       // Сохраняем группу, если хотя бы одна опция прошла фильтрацию
//   //       return item.options.length > 0
//   //     } else {
//   //       // Фильтруем обычную опцию
//   //       return filterOptionsFn(item, filterValue)
//   //     }
//   //   })
//   // })
// }

// const filterOptionsFn = (
//   option: SelectOption,
//   filterValue: string
// ): boolean => {
//   const label =
//     typeof option.label === 'string'
//       ? option.label
//       : reactNodeToString(option.label)
//   return label.toLocaleLowerCase().includes(filterValue.toLocaleLowerCase())
// }

// /**
//  * Получаем из списка опций с вложенными в группы опциями плоский список для более простого рендеринга в jsx
//  */
// export const getFlatOptionsForRender = (
//   options: SelectOptions
// ): Array<SelectOption | { isGroup: true; groupLabel: string }> => {
//   const flatList: Array<SelectOption | { isGroup: true; groupLabel: string }> =
//     []

//   options.forEach((item) => {
//     const isGroup = 'options' in item

//     if (isGroup) {
//       flatList.push(
//         ...[
//           {
//             isGroup: true,
//             groupLabel: item.label
//           },
//           ...item.options
//         ]
//       )
//     } else {
//       flatList.push(item)
//     }
//   })

//   return flatList
// }

// export const getFlatOptions = (options: SelectOptions): SelectOption[] => {
//   const flatList: SelectOption[] = []

//   options.forEach((item) => {
//     const isGroup = 'options' in item

//     if (isGroup) {
//       flatList.push(...item.options)
//     } else {
//       flatList.push(item)
//     }
//   })

//   return flatList
// }

// const checkHasSelectedValue = (value: SelectValue): boolean => {
//   if (Array.isArray(value)) {
//     return value.length > 0
//   }
//   if (typeof value === 'number') {
//     return Number.isInteger(value)
//   }
//   return Boolean(value)
// }

// export const getDisplayValue = (
//   options: SelectOption[],
//   value: SelectValue | SelectValue[]
// ): string => {
//   const hasValue = checkHasSelectedValue(value)

//   if (hasValue) {
//     if (Array.isArray(value)) {
//       return options
//         .filter((option) => value?.includes(option.value))
//         .map((option) => option.label)
//         .join(', ')
//     } else {
//       return String(options.find((option) => option.value === value)?.value)
//     }
//   } else {
//     return ''
//   }

//   // if (multiple) {
//   //   ;``
//   //   displayValue = options
//   //     .filter((option) => value?.includes(option.value))

//   //     .map((i) => i.label)
//   //     .join(', ')
//   // } else {
//   //   displayValue = getItemByValue(value as string | number, options)?.label
//   // }
// }
