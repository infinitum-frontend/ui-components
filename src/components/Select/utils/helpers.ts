import { reactNodeToString } from '~/src/utils/helpers'
import {
  FlattenOption,
  SelectOption,
  SelectOptionGroup,
  SelectOptions
} from './types'

// TODO: поправить типы, убрать @ts-expect-error

export const getFilteredFlattenOptions = (
  options: FlattenOption[],
  filterValue: string
): SelectOption[] => {
  // @ts-expect-error
  return options.filter((option) => {
    if (isGroupLabel(option)) {
      return false
    }
    // @ts-expect-error
    return filterOptionsFn(option, filterValue)
  })
}

export const getFilteredGroupedOptions = (
  options: SelectOptions,
  filterValue: string
): SelectOptions => {
  return options.reduce<SelectOptions>((acc, item) => {
    if ('options' in item) {
      const filteredInGroup = item.options.filter((option: SelectOption) =>
        filterOptionsFn(option, filterValue)
      )
      if (filteredInGroup.length) {
        acc.push({
          label: item.label,
          options: filteredInGroup
        })
      }
    } else if (filterOptionsFn(item, filterValue)) {
      acc.push(item)
    }
    return acc
  }, [])
}

const filterOptionsFn = (
  option: SelectOption,
  filterValue: string
): boolean => {
  const label =
    typeof option.label === 'string'
      ? option.label
      : reactNodeToString(option.label)

  const searchStr = filterValue.toLocaleLowerCase()
  const labelStr = label.toLocaleLowerCase()

  return labelStr.includes(searchStr)
}

export const getFlattenOptions = (options: SelectOptions): FlattenOption[] => {
  return options.flatMap((item) => {
    if ('options' in item) {
      return [{ groupLabel: item.label }, ...item.options]
    }
    return item
  })
}

export const isGroupLabel = (option: FlattenOption): boolean => {
  return 'groupLabel' in option
}

export const isGroupOption = (
  option: SelectOption | SelectOptionGroup
): boolean => {
  return 'options' in option
}
