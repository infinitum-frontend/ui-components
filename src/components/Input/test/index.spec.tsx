import { describe, it, expect, vi } from 'vitest'
import { Input } from '../index'
import { renderComponent } from '@/testSetup'
import { act, fireEvent, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { TextFieldClasses } from '../../../utils/textFieldClasses'

describe('input', () => {
  it('should render', () => {
    const { el } = renderComponent(<Input />)
    expect(el).toBeDefined()
    expect(el.className).toContain('inf-input')
  })

  it('should render as plain input by default', () => {
    const { el } = renderComponent(<Input />)
    expect(el).toBeEmptyDOMElement()
    expect(el.tagName).toBe('INPUT')
  })

  it('should render as complex input by default', () => {
    const { el } = renderComponent(<Input prefix={'prefix'} />)
    const inputEl = el.querySelector('input') as HTMLInputElement
    expect(el).not.toBeEmptyDOMElement()
    expect(el).toContainHTML(inputEl.outerHTML)
  })

  it('matches snapshot', () => {
    const { el } = renderComponent(<Input />)
    expect(el).toMatchSnapshot()
  })
})

describe('className', () => {
  it('should apply custom classname on plain input', () => {
    const { el } = renderComponent(<Input className={'custom-class'} />)
    expect(el.className).toContain('custom-class')
    expect(el.className).toContain('inf-input')
  })

  it('should apply custom classname on complex input', () => {
    const { el } = renderComponent(
      <Input className={'custom-class'} prefix={'prefix'} />
    )
    expect(el.className).toContain('custom-class')
    expect(el.className).toContain('inf-input')
  })
})

describe('style', () => {
  it('should apply style on plain input', () => {
    const { el } = renderComponent(<Input style={{ width: '100%' }} />)
    expect(el).toHaveStyle('width: 100%')
  })

  it('should apply style on complex input', () => {
    const { el } = renderComponent(
      <Input style={{ width: '100%' }} prefix={'prefix'} />
    )
    expect(el).toHaveStyle('width: 100%')
  })
})

describe('size', () => {
  it('should support size on plain input', () => {
    const { el } = renderComponent(<Input />)
    expect(el.className).toContain(TextFieldClasses.size.medium)
  })

  it('should support size on complex input', () => {
    const { el } = renderComponent(<Input prefix={'prefix'} />)
    expect(el)
  })
})

describe('value', () => {
  it('should apply value on plain input', () => {
    const { el } = renderComponent(<Input value={'Акции'} />)
    expect(el).toHaveValue('Акции')
  })

  it('should apply value on complex input', () => {
    const { el } = renderComponent(<Input value={'Акции'} prefix={'prefix'} />)
    const inputEl = el.querySelector('input')
    expect(inputEl).toHaveValue('Акции')
  })
})

describe('events', () => {
  it('should support onInput', () => {
    let eventValue
    let eventType
    const onInput = vi.fn((value, e) => {
      eventValue = value
      eventType = e.type
    })

    const { el } = renderComponent(<Input onInput={onInput} />)
    fireEvent.input(el, { target: { value: 'hello' } })

    expect(onInput).toHaveBeenCalledOnce()
    expect(eventValue).toBe('hello')
    expect(eventType).toBe('change')
  })

  it('should support onBlur', () => {
    let event
    const onBlur = vi.fn((e) => {
      event = e
    })
    const { el } = renderComponent(<Input onBlur={onBlur} />)
    fireEvent.blur(el)

    expect(onBlur).toHaveBeenCalledOnce()
    expect(onBlur).toBeCalledWith(event)
  })

  it('should support onFocus', () => {
    let event
    const onFocus = vi.fn((e) => {
      event = e
    })
    const { el } = renderComponent(<Input onFocus={onFocus} />)
    fireEvent.focus(el)

    expect(onFocus).toHaveBeenCalledOnce()
    expect(onFocus).toHaveBeenCalledWith(event)
  })
})

describe('formatter', () => {
  const defaultValue = 'облигации'
  it('should transform value', () => {
    const { el } = renderComponent(
      <Input formatter={(value) => value?.toUpperCase()} value={defaultValue} />
    )
    expect(el).toHaveValue(defaultValue.toUpperCase())
  })

  it('should emit transformed value', () => {
    let eventValue
    const onInput = vi.fn((value) => {
      eventValue = value
    })
    const { el } = renderComponent(
      <Input formatter={(value) => value?.toUpperCase()} onInput={onInput} />
    )

    fireEvent.input(el, {
      target: { value: defaultValue }
    })

    expect(onInput).toBeCalledTimes(1)
    expect(eventValue).toBe(defaultValue.toUpperCase())
  })
})

describe('placeholder', () => {
  const defaultPlaceholder = 'Введите значение'

  it('should be applied on plain input', () => {
    const { el } = renderComponent(<Input placeholder={defaultPlaceholder} />)
    expect((el as HTMLInputElement).placeholder).toBe(defaultPlaceholder)
  })

  it('should be applied on complex input', () => {
    const { el } = renderComponent(
      <Input placeholder={defaultPlaceholder} prefix={'prefix'} />
    )

    const inputEl = el.querySelector('input') as HTMLInputElement
    expect(inputEl.placeholder).toBe(defaultPlaceholder)
  })

  it('should be empty on plain focused input', () => {
    const { el } = renderComponent(<Input placeholder={defaultPlaceholder} />)

    fireEvent.focus(el)
    expect((el as HTMLInputElement).placeholder).toBe('')
  })

  it('should be empty on complex focused input', () => {
    const { el } = renderComponent(
      <Input placeholder={defaultPlaceholder} prefix={'prefix'} />
    )
    const inputEl = el.querySelector('input') as HTMLInputElement
    fireEvent.click(el)
    expect(inputEl.placeholder).toBe('')
  })
})

describe('border', () => {
  it('should be regular on plain input', async () => {
    const { el } = renderComponent(<Input />)
    expect(el).toHaveClass(TextFieldClasses.borderRadius.regular)
    expect(el).toHaveStyle('border-radius: var(--inf-border-radius-small);')
  })

  it('should be regular on complex input', () => {
    const { el } = renderComponent(<Input prefix={'prefix'} />)
    expect(el).toHaveClass(TextFieldClasses.borderRadius.regular)
    expect(el).toHaveStyle('border-radius: var(--inf-border-radius-small);')
  })

  it('should support noBorder on plain input', () => {
    const { el } = renderComponent(<Input noBorder={true} />)
    expect(el).toHaveClass(TextFieldClasses.noBorder)
    expect(el).toHaveStyle('border: none')
  })

  it('should support noBorder on complex input', () => {
    const { el } = renderComponent(<Input noBorder={true} prefix={'prefix'} />)
    expect(el).toHaveClass(TextFieldClasses.noBorder)
    expect(el).toHaveStyle('border: none')
  })

  it('should support filled class on plain input', () => {
    const { el } = renderComponent(<Input value={'test'} />)
    expect(el).toHaveClass(TextFieldClasses.filled)
  })

  it('should support filled class on complex input', () => {
    // ожидаем, чтобы корректно отработал ref
    act(() => {
      renderComponent(<Input value={'test'} prefix={'prefix'} />)
    })
    const el = screen.queryByRole('textbox')
    expect(el).toHaveClass(TextFieldClasses.filled)
  })
})

describe('disabled', () => {
  it('should apply on plain input', () => {
    const { el } = renderComponent(<Input disabled={true} />)
    expect(el).toHaveProperty('disabled')
  })

  it('should apply on complex input', () => {
    const { el } = renderComponent(<Input disabled={true} prefix={'prefix'} />)
    const inputEl = el.querySelector('input')
    expect(el).toHaveClass(TextFieldClasses.disabled)
    expect(inputEl).toHaveProperty('disabled')
  })
})

describe('status', () => {
  it('should apply on plain input', () => {
    const { el } = renderComponent(<Input status={'error'} />)
    expect(el).toHaveClass(TextFieldClasses.status.error)
    expect(el).toHaveStyle('border-color: var(--inf-color-primary);')
  })

  it('should apply on complex input', () => {
    const { el } = renderComponent(<Input status={'error'} prefix={'prefix'} />)
    expect(el).toHaveClass(TextFieldClasses.status.error)
    expect(el).toHaveStyle('border-color: var(--inf-color-primary);')
  })
})

describe('clear button', () => {
  const user = userEvent.setup()

  it('should be visible', () => {
    const { el } = renderComponent(<Input allowClear={true} />)
    const clearButton = el.querySelector(
      '.inf-input-wrapper__clear-button'
    ) as HTMLSpanElement

    expect(clearButton).toBeInTheDocument()
    expect(clearButton).toBeVisible()
  })

  it('should focus input after click', async () => {
    const { el } = renderComponent(
      <Input allowClear={true} value={'Облигации'} />
    )
    const clearButton = el.querySelector(
      '.inf-input-wrapper__clear-button'
    ) as HTMLSpanElement
    const inputEl = el.querySelector('input')
    await user.click(clearButton)

    expect(el).toHaveClass(TextFieldClasses.focused)
    expect(inputEl).toHaveFocus()
  })

  it('should not work on disabled', async () => {
    const onInput = vi.fn()
    const { el } = renderComponent(
      <Input
        allowClear={true}
        value={'Облигации'}
        onInput={onInput}
        disabled={true}
      />
    )

    const clearButton = el.querySelector(
      '.inf-input-wrapper__clear-button'
    ) as HTMLSpanElement

    await user.click(clearButton).catch((error) => {
      return error
    })
    expect(onInput).toHaveBeenCalledTimes(0)
  })
})
