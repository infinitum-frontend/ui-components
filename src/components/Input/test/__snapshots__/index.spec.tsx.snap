// Vitest Snapshot v1

exports[`input > matches snapshop 1`] = `
<input
  class="inf-input inf-input--size-medium inf-input--br-regular"
  placeholder="Введите значение"
  value=""
/>
`;

exports[`input > matches snapshot 1`] = `
<input
  class="inf-input inf-input--size-medium inf-input--br-regular"
  placeholder="Введите значение"
  value=""
/>
`;
